### PR TITLE
Separate return type tests and create matrix runner

### PR DESCRIPTION
**Pull request by rmoyard**
**Changes**

1. The return type tests are separated for execution on Github action via a matrix. It reduces the risk of crash of the github workers and reduces the time to run the test. 
2. Some unexpected changes were needed for some import of tf and jax.
3. Added back tests that were deleted in https://github.com/PennyLaneAI/pennylane/pull/3235

### CODE REVIEW

File: tests/returntypes/jax/test_jax_qnode_new.py [hunk 1]
```diff
@@ -1391,3 +1388,687 @@ def circuit(data, weights, coeffs):
...
+    ["default.qubit", "adjoint", "forward"],
+    ["default.qubit", "adjoint", "backward"],
+]
+
+jacobian_fn = [jax.jacobian, jax.jacrev, jax.jacfwd]
+
+
+@pytest.mark.parametrize("dev_name,diff_method,mode", qubit_device_and_diff_method_and_mode)
+@pytest.mark.parametrize("shots", [None, 10000])
+class TestReturn:
```
Comment (rmoyard) [inline tests/returntypes/jax/test_jax_qnode_new.py:1406]: This class was deleted in https://github.com/PennyLaneAI/pennylane/pull/3235 but it should not have been the case.

File: tests/returntypes/jax/test_jax_jit_new.py [hunk 2]
```diff
@@ -18,11 +18,8 @@
 pytestmark = pytest.mark.jax
 
 jax = pytest.importorskip("jax")
-jnp = pytest.importorskip("jax.numpy")
-
-from jax.config import config
-
-config.update("jax_enable_x64", True)
```
Comment (antalszava) [inline tests/returntypes/jax/test_jax_jit_new.py:25]: Weird that this had problems :thinking:

File: tests/returntypes/jax/test_jax_qnode_new.py (hunk 1 above)
Comment (antalszava) [reply tests/returntypes/jax/test_jax_qnode_new.py:1406]: Oh boy, thank you for catching this! :pray:

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -202,6 +202,18 @@ jobs:
...
+    strategy:
+      matrix:
+        config:
+          - { loc: "tests/returntypes/jax", suite: 'jax'}
+          - { loc: "tests/returntypes/tf", suite: 'tf'}
+          - { loc: "tests/returntypes/autograd", suite: 'autograd'}
+          - { loc: "tests/returntypes/gradients", suite: 'gradients'}
+          - { loc: "tests/returntypes/transforms", suite: 'transforms'}
+          - { loc: "tests/returntypes/paramshift", suite: 'paramshift'}
+          - { loc: "tests/returntypes/finitediff", suite: 'finitediff'}
```
Comment (antalszava) [inline .github/workflows/tests.yml:214]: I think these two should be moved into `gradients` because they are gradient transform tests.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -202,6 +202,18 @@ jobs:
...
+          - { loc: "tests/returntypes/others", suite: 'others'}
```
Comment (antalszava) [inline .github/workflows/tests.yml:215]: I don't think `others` is a descriptive name enough and it's creating an opaque container of test files - it may serve us in the short term, but I fear it remains as such and will be accepted as the main name. I think we should mirror the `pennylane/module_name` sort of convention in this suite too, unless a deviation is very much understandable (e.g., having `autograd` is simpler and likely more intuitive than creating an `interfaces` folder which then either has an `autograd` subfolder or related test files).

I think we'd better simply leave the files that don't belong in a folder in the root `returntypes` folder.

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @rmoyard, great changes! :clap: This is very much needed in the tests. I think another look at the folder structure could be worth it to make sure it's intuitive for ppl. outside of the project too.

Thank you for catching the removed tests!

File: .github/workflows/tests.yml (hunk 3 above)
Comment (rmoyard) [reply .github/workflows/tests.yml:214]: I did that because the tests are way too long, and the other ones are very fast.

**Review by antalszava — APPROVED** [review]
Looks good! :clap:
